### PR TITLE
Adding ObjectSelector to admission webhooks

### PR DESCRIPTION
**Pull request by caesarxuchao**
/assign
/sig api-machinery

Implementing [KEP#objectSelector](https://github.com/liggitt/enhancements/blob/7b7fc0c2293c02c9bb580805dbdae49d802a189b/keps/sig-api-machinery/00xx-admission-webhooks-to-ga.md#object-selector).

I only implemented it for the main resource. The behavior for subresources is undefined.

I added to the admission webhook integration test to make sure webhooks are called if labels matches the object selector. The test also makes sure in-flight label changes made by the mutation webhooks are not counted when compared to the object selector.

TODO:
1. update the API doc to warn people that the objectSelector should only be used for opt-in webhooks.

I still need to do some code cleanup, so marked as WIP. cc @liggitt @jpbetz @roycaihw 

```release-note
Added object selector to admission webhook configurations.
The object selector applies to an object's labels. For CREATE and UPDATE, the object is the new object. For DELETE, it is existing object (passed as oldObject in the admissionRequest).
Admission behavior for subresources with non-empty object selector is undefined currently.
Use the object selector only if the webhook is opt-in, because end users may skip the admission webhook by setting the labels.
````

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/patch.go [hunk 1]
```diff
@@ -492,16 +496,26 @@ func (p *patcher) applyPatch(_ context.Context, _, currentObject runtime.Object)
...
+	accessor, err := meta.Accessor(patchedObject)
+	if err != nil {
+		return nil, errors.NewInternalError(err)
+	}
+	labelsHolder, ok := request.LabelsHolderFrom(ctx)
+	if !ok {
+		debug.PrintStack()
+		return nil, errors.NewInternalError(fmt.Errorf("expect labelsHolder set in context"))
+	}
+	labelsHolder.SetLabels(accessor.GetLabels())
```
Comment (caesarxuchao) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/patch.go]: The changes to the patch handler are complicated. I need to extract the labels after the patch is applied to the existing object, but before the running the mutation webhooks (`applyAdmission`), and pass the labels to the validation webhooks (encapsuled in `createValidation` and `updateValidation`). I used the `*ctx.LabelsHolder` to pass the labels.

**Comment by caesarxuchao** [comment]
/assign @jpbetz @liggitt 
/unassign

I can still clean up the integration test more, but I want to get some early feedback. Thank you.

**Comment by caesarxuchao** [comment]
Though prow said TestWebhookV1beta1 was failing in the integration test, the failed test might actually be the TestBrokenWebhook, as that's what was reported as broken if I ran `go test ./...` locally.

I'll take another look tomorrow.

**Comment by caesarxuchao** [comment]
I extracted the failure [log](https://pastebin.com/VSBh0Ctp) in the artifact. It claimed the TestWebhookV1beta1 failed, but it seems that all subtests passed..

**Comment by caesarxuchao** [comment]
This was why the test failed in the first run:
```
admission_test.go:1176: objects shouldn't be dispatched to this webhook based on the labels added by the mutation webhook. resource={apps v1 daemonsets}, subresource=, operation=UPDATE
```

This was why the test failed in the second run:
```
admission_test.go:1176: objects shouldn't be dispatched to this webhook based on the labels added by the mutation webhook. resource={apps v1beta2 daemonsets}, subresource=, operation=UPDATE
```

File: test/integration/apiserver/admissionwebhook/admission_test.go [hunk 2]
```diff
@@ -1161,6 +1328,22 @@ func getStubObj(gvr schema.GroupVersionResource, resource metav1.APIResource) (*
...
 
+// cleanLabels clears the labels on existing objects. This operation is not
+// subject to the object selector focused webhooks, as the updated object
+// doesn't have any labels.
+func cleanLabels(client dynamic.Interface, gvr schema.GroupVersionResource, obj *unstructured.Unstructured) (*unstructured.Unstructured, error) {
+	obj.SetLabels(nil)
+	var updateObj *unstructured.Unstructured
+	if err := retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+		var err error
+		updateObj, err = client.Resource(gvr).Namespace(obj.GetNamespace()).Update(obj, metav1.UpdateOptions{})
```
[the remark this answers is not in the document]
Comment (caesarxuchao) [reply test/integration/apiserver/admissionwebhook/admission_test.go]: Thanks. I'll update the code.

**Comment by caesarxuchao** [comment]
I believe that the flake I encountered was due to the watch cache. The object in watch cache contains already removed labels, thus the request was dispatched to wrong webhooks.

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/create.go [hunk 3]
```diff
@@ -129,7 +130,13 @@ func createHandler(r rest.NamedCreater, scope *RequestScope, admit admission.Int
 		audit.LogRequestObject(ae, obj, scope.Resource, scope.Subresource, scope.Serializer)
 
 		userInfo, _ := request.UserFrom(ctx)
-		admissionAttributes := admission.NewAttributesRecord(obj, nil, scope.Kind, namespace, name, scope.Resource, scope.Subresource, admission.Create, options, dryrun.IsDryRun(options.DryRun), userInfo)
+
+		var labels map[string]string
+		accessor, err := meta.Accessor(obj)
+		if err == nil {
+			labels = accessor.GetLabels()
+		}
```
[the remark this answers is not in the document]
Comment (caesarxuchao) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/create.go]: extensions/v1beta1/deployments/rollback does not implement the Object interface. That seems to be the only exceptions.

I updated the update.go to return NewInternalError if meta.Accessor() fails.

**Comment by caesarxuchao** [comment]
/retest

Integration test failure not related.

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/patch.go [hunk 4]
```diff
@@ -492,16 +495,26 @@ func (p *patcher) applyPatch(_ context.Context, _, currentObject runtime.Object)
...
+	accessor, err := meta.Accessor(patchedObject)
+	if err != nil {
+		return nil, errors.NewInternalError(err)
+	}
+	labels := accessor.GetLabels()
+	labelsHolder, ok := request.LabelsHolderFrom(ctx)
+	if !ok {
+		return nil, errors.NewInternalError(fmt.Errorf("expect labelsHolder set in context"))
+	}
+	labelsHolder.SetLabels(labels)
```
Comment (jpbetz) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/patch.go]: Add comment here summarizing why/how labels are messaged via the context? Seems like we should comment about that somewhere..

File: pkg/apis/admissionregistration/types.go [hunk 5]
```diff
@@ -223,6 +223,17 @@ type Webhook struct {
...
+	// The object selector applies to an object's labels. For CREATE and
+	// UPDATE, the object is the new object. For DELETE, it is existing
+	// object (passed as oldObject in the admissionRequest).
+	// Admission behavior for subresources with non-empty object selector
+	// is undefined currently.
+	// Use the object selector only if the webhook is opt-in, because end
+	// users may skip the admission webhook by setting the labels.
+	// Default to the empty LabelSelector, which matches everything.
+	// +optional
+	ObjectSelector *metav1.LabelSelector
```
Comment (sttts) [inline pkg/apis/admissionregistration/types.go]: Why not `ObjectLabelSelector` ?

File: pkg/apis/admissionregistration/types.go [hunk 6]
```diff
@@ -223,6 +223,17 @@ type Webhook struct {
 	// +optional
 	NamespaceSelector *metav1.LabelSelector
 
+	// The object selector applies to an object's labels. For CREATE and
+	// UPDATE, the object is the new object. For DELETE, it is existing
+	// object (passed as oldObject in the admissionRequest).
+	// Admission behavior for subresources with non-empty object selector
+	// is undefined currently.
```
Comment (sttts) [inline pkg/apis/admissionregistration/types.go]: can we define it?

File: staging/src/k8s.io/apiserver/pkg/admission/attributes.go [hunk 7]
```diff
@@ -58,6 +59,7 @@ func NewAttributesRecord(object runtime.Object, oldObject runtime.Object, kind s
 		dryRun:      dryRun,
 		object:      object,
 		oldObject:   oldObject,
+		labels:      labels,
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/admission/attributes.go]: why do we need labels globally throughout the code base and not only in the webhooks. This looks wrong. Please move that logic into the webhook.

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/patch.go [hunk 8]
```diff
@@ -242,7 +245,7 @@ func PatchResource(r rest.Patcher, scope *RequestScope, admit admission.Interfac
 	}
 }
 
-type mutateObjectUpdateFunc func(obj, old runtime.Object) error
+type mutateObjectUpdateFunc func(ctx context.Context, obj, old runtime.Object) error
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/patch.go]: why this change?

**Comment by sttts** [comment]
I am not happy about the ripples of the change. I am against merging it in this way.

/hold

**Comment by caesarxuchao** [comment]
The [requirement](https://github.com/liggitt/enhancements/blob/7b7fc0c2293c02c9bb580805dbdae49d802a189b/keps/sig-api-machinery/00xx-admission-webhooks-to-ga.md#object-selector) of extracting labels "at the admission time before running any plugins and will be used for all webhooks during the admission process" makes the implementation complicated. Especially for PATCH operations, I needed to extract the labels after the patch is applied, but before applying the mutation webhooks in this [line](https://github.com/kubernetes/kubernetes/blob/afba9c035510c08378cac5bfb3ab4dcb72577129/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/patch.go#L558), thus I used `context` to hold and pass the labels.

One alternative I had considered was letting the [mutation webhook controller](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go#L58) extract the labels. The drawbacks are:
1. we are assuming that all built-in admission plugins are not and will not change object labels.
2. it doesn't work for subresources. According to the [KEP](https://github.com/liggitt/enhancements/blob/7b7fc0c2293c02c9bb580805dbdae49d802a189b/keps/sig-api-machinery/00xx-admission-webhooks-to-ga.md#object-selector), sometimes we need to extract the labels from the parent object, which is not accessible by the [mutation webhook controller](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go#L58). 

@sttts

**Comment by liggitt** [comment]
Talked through the various issues with @sttts, @caesarxuchao, @jpbetz, and @roycaihw. most of the difficulties with this implementation came from the original behavior specified in the KEP, which in retrospect led to very counter-intuitive behavior from an end-user's perspective.

The updated behavior we landed on (which is much easier to explain and to implement... double win):

To determine whether the objectSelector matches given admission attributes:
* An empty objectSelector matches all requests
* A non-empty objectSelector is evaluated against both the oldObject and newObject that would be sent to the webhook, and is considered to match if either object matches the selector. A null object or an object that cannot have labels is not considered to match.

**Comment by sttts** [comment]
I very much prefer what @liggitt suggests in https://github.com/kubernetes/kubernetes/pull/78226#issuecomment-497102966. Feel free to unhold the PR with such an implementation.

**Comment by caesarxuchao** [comment]
Superseded by #78505
